internal/agentdiscovery/agentdiscovery.go

```diff
@@ -0,0 +1,159 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2022 Datadog, Inc.
+
+// Package agentdiscovery provides a mechanism for automatically discovering a
+// Datadog agent and its available features
+package agentdiscovery
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"time"
+)
+
+const (
+	// defaultSocketAPM specifies the socket path to use for connecting to the trace-agent.
+	defaultSocketAPM   = "/var/run/datadog/apm.socket"
+	defaultHostname    = "localhost"
+	defaultPort        = "8126"
+	defaultHTTPTimeout = 2 * time.Second
+)
+
+// Features is functionality reported as available by the trace-agent
+type Features struct {
+	// Endpoints are the available data ingestion endpoints such as /traces,
+	// /profiles, etc.
+	Endpoints []string `json:"endpoints"`
+	// ClientDropP0s indicates whether the tracer client should [ ... SOMETHING ... ]
+	ClientDropP0s bool `json:"client_drop_p0s"`
+	// StatsdPort is the UDP port where the agent accepts DogStatsD metrics
+	StatsdPort int `json:"statsd_port"`
+	// FeatureFlags are enabled agent feature flags
+	FeatureFlags []string `json:"feature_flags"`
+}
+
+// AgentFeatures queries the agent's /info endpoint
+func AgentFeatures(addr string, client *http.Client) (*Features, error) {
+	u := url.URL{
+		Scheme: "http",
+		Host:   addr,
+		Path:   "/info",
+	}
+	fmt.Fprintf(os.Stderr, "trying URL %s\n", u.String())
+	resp, err := client.Get(u.String())
+	if err != nil {
+		return nil, fmt.Errorf("loading features: %w", err)
+	}
+	if resp.StatusCode == http.StatusNotFound {
+		// agent is older than 7.28.0, features not discoverable
+		return nil, errors.New("agent does not support /info endpoint")
+	}
+	defer resp.Body.Close()
+	var info Features
+	if err := json.NewDecoder(resp.Body).Decode(&info); err != nil {
+		return nil, fmt.Errorf("decoding features: %w", err)
+	}
+	return &info, nil
+}
+
+var (
+	defaultDialer = &net.Dialer{
+		Timeout:   30 * time.Second,
+		KeepAlive: 30 * time.Second,
+		DualStack: true,
+	}
+
+	defaultClient = &http.Client{
+		// We copy the transport to avoid using the default one, as it might be
+		// augmented with tracing and we don't want these calls to be recorded.
+		// See https://golang.org/pkg/net/http/#DefaultTransport .
+		Transport: &http.Transport{
+			Proxy:                 http.ProxyFromEnvironment,
+			DialContext:           defaultDialer.DialContext,
+			MaxIdleConns:          100,
+			IdleConnTimeout:       90 * time.Second,
+			TLSHandshakeTimeout:   10 * time.Second,
+			ExpectContinueTimeout: 1 * time.Second,
+		},
+		Timeout: defaultHTTPTimeout,
+	}
+)
+
+// HTTPClient returns the default http.Client to start the tracer with.
```

Comment:
Is this HTTPClient Tracer-specific? Can other packages use it to talk to the agent? 
Let's make this as generic as possible.

Reply:
This was a copy-paste from the tracer. I've updated the comment, but yes, this client is intended to be used by the tracer, profiler, or anything in the future as a default client.